### PR TITLE
Replace sync/atomic with uber-go/atomic

### PR DESCRIPTION
**Pull request by jvrplmlmn**
Following https://github.com/prometheus/prometheus/pull/7659, this addresses https://github.com/prometheus/prometheus/issues/7647.

PTAL @codesome 🙏

### CODE REVIEW

File: Makefile.common [hunk 1]
```diff
@@ -118,7 +118,7 @@ endif
 %: common-% ;
 
 .PHONY: common-all
-common-all: precheck style check_license lint unused build test
+common-all: precheck restricted_imports style check_license lint unused build test
```
Comment (jvrplmlmn) [inline Makefile.common:121]: I am still missing a lot of context, so I am not completely sure that this is the best place to put this, suggestions definitely welcome.

File: cmd/prometheus/main.go [hunk 2]
```diff
@@ -67,6 +66,8 @@ import (
 	"github.com/prometheus/prometheus/tsdb"
 	"github.com/prometheus/prometheus/util/strutil"
 	"github.com/prometheus/prometheus/web"
+
+	"go.uber.org/atomic"
```
Comment (codesome) [inline cmd/prometheus/main.go:70]: This should be with the other non `prometheus/prometheus` imports above (same for remaining files)

File: notifier/notifier_test.go [hunk 3]
```diff
@@ -102,10 +103,12 @@ func TestHandlerSendAll(t *testing.T) {
...
 		errc             = make(chan error, 1)
 		expected         = make([]*Alert, 0, maxBatchSize)
-		status1, status2 = int32(http.StatusOK), int32(http.StatusOK)
+		status1, status2 atomic.Int32
 	)
+	status1.Store(int32(http.StatusOK))
+	status2.Store(int32(http.StatusOK))
 
-	newHTTPServer := func(u, p string, status *int32) *httptest.Server {
+	newHTTPServer := func(u, p string, status atomic.Int32) *httptest.Server {
```
Comment (codesome) [inline notifier/notifier_test.go:111]:
```suggestion
	newHTTPServer := func(u, p string, status *atomic.Int32) *httptest.Server {
```

File: notifier/notifier_test.go [hunk 4]
```diff
@@ -194,11 +197,11 @@ func TestHandlerSendAll(t *testing.T) {
 	testutil.Assert(t, h.sendAll(h.queue...), "all sends failed unexpectedly")
 	checkNoErr()
 
-	atomic.StoreInt32(&status1, int32(http.StatusNotFound))
+	status1.Store(int32(http.StatusNotFound))
 	testutil.Assert(t, h.sendAll(h.queue...), "all sends failed unexpectedly")
 	checkNoErr()
 
-	atomic.StoreInt32(&status2, int32(http.StatusInternalServerError))
+	status2.Store(int32(http.StatusNotFound))
```
Comment (codesome) [inline notifier/notifier_test.go:204]: Status code changed by mistake I believe

File: storage/remote/intern.go [hunk 5]
```diff
@@ -40,13 +40,17 @@ type pool struct {
...
-	// See https://golang.org/pkg/sync/atomic/#pkg-note-BUG for more info.
-	refs int64
+	refs atomic.Int64
 
 	s string
 }
 
+func newEntry(s string) *entry {
+	entry := &entry{s: s}
+	entry.refs.Store(1)
```
Comment (codesome) [inline storage/remote/intern.go:50]: I would expect a `newEntry` to have 0 refs in the returned object. So let's skip storing 1 here (while keeping the method) and increment where `newEntry` is called.

File: cmd/prometheus/main.go (hunk 2 above)
Comment (jvrplmlmn) [reply cmd/prometheus/main.go:70]: Sure 👍  

Is this verified by `goimports` on CI?

File: storage/remote/intern.go (hunk 5 above)
Comment (jvrplmlmn) [reply storage/remote/intern.go]: Addressed in 9ef9045

File: notifier/notifier_test.go (hunk 4 above)
Comment (jvrplmlmn) [reply notifier/notifier_test.go]: Good catch! Addressed in 1b3f578

File: notifier/notifier_test.go (hunk 3 above)
Comment (jvrplmlmn) [reply notifier/notifier_test.go]: Addressed in 1b3f578

**Comment by jvrplmlmn** [comment]
Thanks for the review @codesome, I have addressed your points, hope its ready to ship now.

**Comment by roidelapluie** [comment]
Can you make the ci changes in another pr?

**Comment by jvrplmlmn** [comment]
@roidelapluie sure thing, lets keep 1 change per PR.

**Comment by roidelapluie** [comment]
> @roidelapluie sure thing, lets keep 1 change per PR.

This is because the Makefile change per se it more complex and I don't want to block the atomic/ part.

**Comment by jvrplmlmn** [comment]
Let me know if you would like me to rebase the branch to get rid of the commits that address the Makefile/CI part.

**Comment by roidelapluie** [comment]
We will probably just squash the PR in github

File: web/web.go [hunk 6]
```diff
@@ -51,23 +50,23 @@ import (
...
+	"github.com/prometheus/prometheus/tsdb"
+	"github.com/prometheus/prometheus/tsdb/index"
 	"github.com/prometheus/prometheus/util/httputil"
 	api_v1 "github.com/prometheus/prometheus/web/api/v1"
 	api_v2 "github.com/prometheus/prometheus/web/api/v2"
 	"github.com/prometheus/prometheus/web/ui"
+	"github.com/soheilhy/cmux"
+	"go.uber.org/atomic"
+	"golang.org/x/net/netutil"
+	"google.golang.org/grpc"
```
Comment (codesome) [inline web/web.go:69]: Not `prometheus/prometheus` :) so let's keep it above and separated for consistency

File: cmd/prometheus/main.go (hunk 2 above)
Comment (codesome) [reply cmd/prometheus/main.go]: > Is this verified by goimports on CI?

goimports would just sort the imports placed together AFAIK. Grouping relevant imports is on us.

**Comment by codesome** [comment]
I wonder why did unit tests no run. Hoping after fixing the nit it will run again.

File: web/web.go (hunk 6 above)
Comment (jvrplmlmn) [reply web/web.go]: Hopefully finally addressed in 568bd51 🙏

**Pull request merged**